### ISSUE TITLE
[Help] Struggling to set content based on dimensions

### ISSUE THREAD

**Issue by andrewjensen** [gyscos/cursive#453]
First: is this a bug report? A suggestion? Or asking for help?

Asking for help

# Problem description

I am trying to get the dimensions of a view, so I can generate content based on its dimensions. It looks like the correct way to do this is with a `LastSizeView`, so when it gets its layout, its child should be able to access its dimensions and react accordingly.

I'm getting stuck, and I think it's because I don't have a good understanding of how views are initialized. Here is a small program demonstrating what I'm trying to do, with two commented attempts to update the text of a TextView, based on the dimensions stored in the LastSizeView that wraps it:

```rust
use cursive::view::{Identifiable};
use cursive::views::{TextView, ResizedView, LastSizeView};

fn main() {
    let mut siv = cursive::default();

    siv.add_fullscreen_layer(
        ResizedView::with_full_screen(
            LastSizeView::new(
                TextView::new("INITIAL").with_name("main_content")
            ).with_name("tracked_size")
        )
    );

    // Attempt #1 (results in crash)
    // let tracked_size_view = siv.find_name::<LastSizeView<TextView>>("tracked_size").unwrap();
    // let updated_content = get_content(tracked_size_view.size.x, tracked_size_view.size.y);
    // let mut main_content_view = siv.find_name::<TextView>("main_content").unwrap();
    // main_content_view.set_content(updated_content);

    // Attempt #2: (results in seeing "...", not running the second callback)
    // let mut updated_content = "...".to_string();
    // siv.call_on_name("tracked_size", |v: &mut LastSizeView<TextView>| {
    //     updated_content = get_content(v.size.x, v.size.y);
    // });
    // siv.call_on_name("main_content", |v: &mut TextView| {
    //     v.set_content(updated_content);
    // });

    // Expected behavior: A single grey TextView should fill the screen, with text like the following:
    // "This view fills the screen and reports its dimensions.\n\nwidth: 120 height: 50"

    siv.run();
}

fn get_content(view_width: usize, view_height: usize) -> String {
    format!("This view fills the screen and reports its dimensions.\n\nwidth: {} height: {}", view_width, view_height)
}
```

It looks like I could potentially create a custom view and use the `Printer` in its `draw` method, but I want to be able to do this with built-in views, because my work in progress TUI needs to adapt to view dimensions with content inside of a ScrollView. For that use case, I will need to know the available width before generating the text for each list item.

What is the correct way to generate content based on dimensions, like I'm attempting above?

Once I understand this better and have a working example, I would be happy to contribute it to the `examples` folder in a PR! Thanks for a great library.

# Environment

* Operating system: macOS
* Backend used: ncurses
* Current locale: `en_US.UTF-8`
* Cursive version: git, master branch (commit `fc59ee6f1757095e4aaeebe6a1cca47d211db8fe`)

**Comment by gyscos** [comment]
Thanks for the report!

Let's clarify your use-case:

You want to display a list of items, in a vertically-scrollable ScrollView, and I guess you want to justify or right-align the content so it must depend on the width of the view.
If it's something as general as that, then the real solution would be to make it easier to achieve with a built-in `View` without bothering about the sizes.

In the general case, a view can react to its dimensions in two places:
* `View::layout`, where it should compute its content/layout based on the allocated space.
* `View::draw`, where it can align/format the content based on the available space.

In practice these two places are usually equivalent: they are called with the same size. A small difference is that `layout` takes a `&mut self` while `draw` takes a `&self`, but interior mutability (like `RefCell`) can bypass that.
A bigger difference is that potentially, `layout` can be skipped when we're sure nothing changed. `View::needs_relayout` can be used by a view to notify its parent that nothing changed, and therefore `View::layout` might be skipped. `View::draw` is still called at each refresh.
The real difference, though, is in the intent: `View::layout` is designed to be used to compute persistent changes, while `View::draw` is supposed to just adapt/format the pre-computed content. So where you'll want to run logic depends on what you want to do with the size: actually change the content (the text being written, possibly the content of nested views), or just change the alignment.

Now, how do we run logic in this `View::layout` step? It's obvious if you write your own view, but if you want to work with built-in views, then it's less clear how to override this method.
The usual solution is by wrapping the built-in view in a custom view that only intercepts the method you're interested in, and forwards everything else. The [`ViewWrapper`](https://docs.rs/cursive/0.14.0/cursive/view/trait.ViewWrapper.html) trait can be a helpful tool to quickly write such a view.
For quick iteration though, there's an even faster way to do that: [`Canvas`](https://docs.rs/cursive/0.14.0/cursive/views/struct.Canvas.html) is a `View` that lets you override its methods with closures. There's a [`Canvas::wrap`](https://docs.rs/cursive/0.14.0/cursive/views/struct.Canvas.html#method.wrap) method that basically lets you wrap a view and then re-define some methods, without defining any new type:

```rust
let view = TextView::new("foo");
let canvas = Canvas::wrap(view)
    .with_layout(|view, size| {
        // You can run any logic in the layout phase here.

        // Update the content?
        view.set_content(format!("{:?}", size));

        // Forward the layout to the inner view, just in case it needs it.
        view.layout(size);
    })
    .with_draw(|view, printer| {
        // And/or you can also use the draw phase.
        // Right-align text
        // (Here we assume single-line and no markup in the text,
        // a more robust solution would be a bit longer)
        printer.print(
            (0,0),
            &format!(
                "{:>width$}",
                view.get_content().source(),
                width=printer.size.x
            )
        );
    });
```

>  want to be able to do this with built-in views, because my work in progress TUI needs to adapt to view dimensions with content inside of a ScrollView.

Can you precise the issue? In most cases, using a `ScrollView` should be transparent, and work automatically with any custom view.

**Comment by andrewjensen** [comment]
Thank you for the very detailed and thoughtful response! Based on the way `Canvas` can wrap other views, and how it has mutable access to the child view during both `layout` and `draw` phases, it looks like I can use it for my use case.

I'm currently writing a TUI to manage an Apache Kafka cluster, and each "screen" displays tables of items, where each one leads to another screen. For example, the summary screen includes a selectable table of topics, which leads to a topic screen that includes a selectable table of consumer groups, etc. I want each table to use all of its available width, because the names of these topics and 

After I read your message, I realized that I mentioned the wrong `cursive` view: I need a `SelectView` to be aware of its size, not a `ScrollView`. (Actually, since I'm working with long lists of items, I am wrapping a `SelectView` inside of a `ScrollView`.) But based on your advice, I was able to put together the following working demo. Notice how it effectively creates a table of selectable "widget" elements, where the name is left-aligned and the amount is right-aligned. This is possible because the canvas re-renders the `SelectView` labels whenever a layout is triggered. I should be able to use similar logic for the tables in my app.

Screenshot:

<img width="1184" alt="image" src="https://user-images.githubusercontent.com/3885955/81488725-cb478d80-9229-11ea-8537-9cf35c8c8f4b.png">

Code:

```rust
use cursive::utils::span::SpannedString;
use cursive::view::View;
use cursive::views::{Canvas, ResizedView, SelectView};
use std::rc::Rc;

struct Widget {
    id: usize,
    name: String,
    amount: usize,
}

fn main() {
    let widgets = vec![
        Widget {
            id: 1,
            name: "my first widget".to_string(),
            amount: 17461,
        },
        Widget {
            id: 2,
            name: "my second widget".to_string(),
            amount: 952,
        },
        Widget {
            id: 3,
            name: "my third widget".to_string(),
            amount: 1853,
        },
    ];

    let model = Rc::new(widgets);
    let model_ref = Rc::clone(&model);

    let mut siv = cursive::default();

    let view = SelectView::new().with_all(model.iter().map(|widget| {
        // Initialize labels to empty string because we have not yet called layout
        ("", widget.id)
    }));

    let size_aware_select =
        Canvas::wrap(view).with_layout(move |view, size| {
            let available_width = size.x;

            // Update the label for each item, based on the available width
            for idx in 0..view.len() {
                let (widget_label, widget_id) =
                    view.get_item_mut(idx).unwrap();

                // TODO: is there a way to conditionally skip this computation? Or is this closure only called when needed?
                let widget =
                    model_ref.iter().find(|w| w.id == *widget_id).unwrap();
                let mut new_label = SpannedString::new();
                new_label
                    .append(print_widget_list_item(widget, available_width));

                *widget_label = new_label;
            }

            // Forward the layout to the inner view, just in case it needs it.
            view.layout(size);
        });

    siv.add_fullscreen_layer(ResizedView::with_full_screen(size_aware_select));

    siv.run();
}

fn print_widget_list_item(widget: &Widget, available_width: usize) -> String {
    let amount_len = widget.amount.to_string().chars().count();

    format!(
        "{:padded_name_len$}{}",
        widget.name,
        widget.amount,
        padded_name_len = available_width - amount_len
    )
}
```

As a follow-up question, would you say this example is idiomatic and performant? Or is there a better way to implement the selectable table?

If this example is useful, I'd be happy to put together a PR. Let me know!

**Comment by gyscos** [comment]
Small details:

* Ideally here `SelectView` would be flexible enough to use arbitrary views as elements rather than just text; this way you could put there a custom view that would handle alignment for you. We're not here yet though.
* You can create "plain" strings with `StyledString::plain()`, or even just `some_string.into()` (we have `impl From<T: Into<String>> for StyledString`):
```rust
    *widget_label = print_widget_list_item(widget, available_width).into();
```
* Rather than having 2 references to the list, you could either directly store the `Widget`s in the `SelectView`:
```rust
    let view = SelectView::new().with_all(model.iter().map(|widget| {
        // Initialize labels to empty string because we have not yet called layout
        ("", widget)
    }));
```
This way you directly have access to the widget, without having to lookup its ID.
Alternatively, you can use `zip` with the widget list, or even use a hashmap keyed by ID rather than a vector.
I also just pushed `SelectView::iter_mut` to make it easier to iterate on the elements/labels and update them.
* The visible width of text should use `.width()` from the [`UnicodeWidthStr`](https://docs.rs/unicode-width/0.1.7/unicode_width/trait.UnicodeWidthStr.html) trait rather than counting characters. This way it works fine with compound unicode characters or with CJK wide characters (like asian alphabets).
* `ResizedView::with_full_screen` can also be achieved with `.full_screen()` from the [`Resizable`](https://docs.rs/cursive/0.14.0/cursive/view/trait.Resizable.html#method.full_screen) trait.
* Rather than counting the number of digits through an intermediate string (allocation!), it may be cheaper to instead count the width of the label, and then format-align the numbers.:
```rust
format!(
    "{}{:amount_len$}",
    widget.name,
    widget.amount,
    amount_len = available_width - widget.name.width()
);
```

**Comment by andrewjensen** [comment]
Thanks for all the excellent suggestions! I'll make sure to incorporate them into my project. And thank you for adding `SelectView::iter_mut` as well!

As for `cursive_table_view`, I remember looking at it just a little as I first started into my project, and it looked too advanced for what I needed, but I spent some more time looking at the docs this morning, and it looks like it does what I need with a pretty clear API. I'll take another look and maybe I'll switch to it.

I'll go ahead and close the issue since this should solve my use case, whether I go with `cursive_table_view` or build my own solution. I edited the issue title to hopefully make it more discoverable.
